### ISSUE TITLE
Pass host's open files ulimit inside a container

### ISSUE THREAD

**Issue by networkop** [srl-labs/containerlab#824]
The goal is to align max open file limit in the host OS and inside a container.

currently, if host's ulimit is low (e.g. 1024) and container starts with a higher number (e.g. ceos has 1073741816) this may lead to bugs in certain python scripts. For example: 

```bash
$ ulimit -n
1024
$ docker exec -it ceos bash
< inside the container now>
bash-4.2# cat test.sh
#!/usr/bin/python3
import resource
import os

max_fd = resource.getrlimit(resource.RLIMIT_NOFILE)[0]
print(max_fd)
os.closerange(3, max_fd)

bash-4.2# ulimit -n
1073741816

bash-4.2# ./test.sh
1073741816

^Z
[1]+  Stopped                 ./test.sh
bash-4.2# bg
[1]+ ./test.sh &
bash-4.2# top -n1 | grep test
   2400 root      20   0   27504   7192   4500 R 100.0   0.1   0:18.87 test.sh
```

The above python script will block and eat 100% of available CPU core.

**Comment by LimeHat** [comment]
What's the output of "olimit -Hn" on this system?
I believe containers inherit the hard limit (instead of the soft one)

**Comment by LimeHat** [comment]
Or, looking at the docker docs, it might be that it is inherited from the default docker daemon setting instead. Yeah, either way, should be easy to match both.

**Comment by nleiva** [comment]
For what it's worth I see this as well, for example:

```bash
⇨  docker exec -it clab-netgo-srl bash
[root@srl /]# ulimit -n
1000
```
```bash
⇨  docker exec -it clab-netgo-ceos bash
bash-4.2# ulimit -n
1073741816
```
```bash
⇨  ulimit -Hn
524288
```

I'll update Docker to see if this behavior changes. 

```bash
 ⇨  docker version
Client: Docker Engine - Community
 Version:           20.10.13
 API version:       1.41
 Go version:        go1.16.15
 Git commit:        a224086
 Built:             Thu Mar 10 14:08:18 2022
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true
```

**Comment by LimeHat** [comment]
Thanks @nleiva 
Would you mind checking one more thing

1. docker container inspect **_[CONTAINER_NAME]_** -f {{.State.Pid}}
2. grep 'files' /proc/**_[PID]_**/limits

for both containers?

Also, what's the host OS & version?

**Comment by networkop** [comment]
@LimeHat it does look like docker passes its own limits to spawned containers:
```
fedora@testbed ~ ⇨  ulimit -Hn
524288
fedora@testbed ~ ⇨  ulimit -Sn
1024
fedora@testbed ~ ⇨  grep 'files' /proc/$(pgrep -f dockerd)/limits
Max open files            1073741816           1073741816           files

# launched a "test" container

fedora@testbed ~ ⇨  grep 'files' /proc/$(docker container inspect test  -f {{.State.Pid}})/limits
Max open files            1073741816           1073741816           files
```

But that still means that, given the above mismatch, we are open to bugs. 

The way I see it is we could raise the ulimit on hostOS based on dockerD's ulimit or lower them in containers best on host's ulimits.

What do you think is the best way forward?

**Comment by LimeHat** [comment]
Ahhh i see. This is one of those newer fedoras with bumped up defaults. Well. I totally forgot about those with my legacy systems.. 😄 

> But that still means that, given the above mismatch, we are open to bugs.
> The way I see it is we could raise the ulimit on hostOS based on dockerD's ulimit or lower them in containers best on host's ulimits.

I mean, this is technically not a mismatch. The number you see is the actual system limit in this case; and 524288/1024 are currently allocated hard/soft limits to the shell process. Containers in this case aren't children of the shell, so these numbers are independent (sort of). And docker just goes for the maximum (which was capped by lower numbers in earlier kernel/systemd combos).

Anyhow. Practical solution (given that we want to provide seamless, zero-configuration experience to our users..) would be to cap the limits with a reasonable, lower value (e.g. 2^20) while spawning containers.

Can you quickly check if your script works fine if you set the limit to 1048576? (add _**--ulimit nofile=1048576:1048576**_ as a parameter when you launch the test container via docker run)

**Comment by networkop** [comment]
I think it's a little more complicated than that, but I may be just misunderstanding the entire theory about ulimits completely.
Here's what I observe (on fedora 35 running in AWS):
```
fedora@testbed ~ ⇨  ulimit -Hn
524288
fedora@testbed ~ ⇨  ulimit -Sn
1024
fedora@testbed ~ ⇨  grep files /proc/$(pgrep dockerd)/limits
Max open files            1073741816           1073741816           files
fedora@testbed ~ ⇨  cat  /proc/sys/fs/file-max
9223372036854775807
```
So assuming that my shell's ulimits are independent of dockerd's ulimits, my containers should be constrained by 1073741816.
Here's what  I do to test:
```
fedora@testbed ~ ⇨  docker run --rm -it --name test ubuntu bash
root@8389cc2b4a35:/# ulimit -Hn
1073741816
root@8389cc2b4a35:/# ulimit -Sn
1073741816
root@8389cc2b4a35:/# ./test.sh
1073741816
< hangs until killed > 
```

but the following works (note that  594290  >> 524288):

```
fedora@testbed ~ ⇨ docker run --rm -it --name test --ulimit nofile=594290 ubuntu bash
root@d2beed74110c:/# ulimit -Hn
524288
root@d2beed74110c:/# ulimit -Sn
524288
root@d2beed74110c:/# ./test.sh
524288
< exits successfully > 
```
So it looks like there's some sort of limit but it's between the minimal value that user shell gets and the high value that dockerd gets. 

do you guys have any ideas? @LimeHat @nleiva

**Comment by networkop** [comment]
just to confirm there are no overrides in `/etc/security/limits`
```
fedora@testbed ~ ⇨  grep -R file /etc/security/limit*
/etc/security/limits.conf:#This file sets the resource limits for the users logged in via PAM.
/etc/security/limits.conf:#Also note that configuration files in /etc/security/limits.d directory,
/etc/security/limits.conf:#file in case the domain is the same or more specific.
/etc/security/limits.conf:#can be overridden with a wildcard setting in a config file in the
/etc/security/limits.conf:#        - core - limits the core file size (KB)
/etc/security/limits.conf:#        - fsize - maximum filesize (KB)
/etc/security/limits.conf:#        - nofile - max number of open file descriptors
/etc/security/limits.conf:#        - locks - max number of file locks the user can hold
/etc/security/limits.conf:# End of file
```

**Comment by networkop** [comment]
here's an interesting piece: 
```
fedora@testbed ~ ⇨  cat /usr/lib/systemd/system/docker.service | grep -i nofile
LimitNOFILE=infinity
```

**Comment by hellt** [comment]
Hi all
Just wondering what are those real use cases that suffer from the files limit discrepancy?

**Comment by LimeHat** [comment]
@networkop I don't think there's any controversy between what i wrote and what you're seeing? 
2^30 (give or take a few) is the max, docker goes for it.

You might be asking why in this case your script hangs - that i don't know (just suboptimal implementation on huge numbers or some sort of collision when it reaches those) - but doesn't really matter imo. 

Anyway I'll submit a patch to give containers 2^20 and that should do it. No reason to give them more than that.

**Comment by networkop** [comment]
to give a bit of a context, I've come across this issue here: https://github.com/amoffat/sh/blob/c65d81a996441c4bf880626ee2d8cef253392fd2/sh.py#L1956

This is a part of Cumulus Linux's dependencies and leads to one of CL's processes hanging forever.  This issue only manifests itself on Fedora, I haven't seen it happening on Debian/Ubuntu derivatives. 

@LimeHat the problem is that the high enough number doesn't guarantee that your script won't hang. I don't think it's a clab's issue, more of a docker|fedora issue but I was just hoping to understand what is the root cause. Why do you think that 2^30 is good enough? 

@nleiva any chance you can find out why this happens in fedora and not debian/ubuntu?

**Comment by LimeHat** [comment]
No, what I'm saying is that 2^30 causes your issue. 
Indeed it is not a clab issue, however a practical approach is to enforce this value is not used with clab (instead of asking our users to change systemd/kernel/docker settings; or vendors to enforce lower values in their images, as we do in srlinux). 
I'm somewhat confident that 2^20 should be good since it is the old default, but you can try it to confirm as well (I inquired about this above). But you did test just over 2^19 so it should be fine.

If you want to investigate this - I would suggest to start without containers at all, just take Python (preferably the same version as you have inside the container); bump the limits to the max, see what happens; etc.

**Comment by hellt** [comment]
@networkop 
do you want to test what @LimeHat added in #825?
this will get you a binary with the fix
```
docker run --rm -v $(pwd):/workspace ghcr.io/oras-project/oras:v0.12.0 pull ttl.sh/clab-39f8654c:1d
```

**Comment by networkop** [comment]
@hellt @LimeHat thanks guys, that workaround has done the trick. I can see the ulimits set on the container and I can't reproduce the original bug:
```
docker inspect test | jq '.[0].HostConfig.Ulimits'
[
  {
    "Name": "nofile",
    "Hard": 524288,
    "Soft": 524288
  }
]
fedora@testbed ~ ⇨  ulimit -Hn
524288
```

I still don't understand how/why it happened in the first place, but this is fine. maybe @nleiva can ask one of his colleagues for some documentation/explanation?

**Comment by hellt** [comment]
cool, thanks for confirming @networkop 
 merging the PR which will auto-close this issue

**Comment by LimeHat** [comment]
@networkop 
> I still don't understand how/why it happened in the first place

My main guess is still a suboptimal implementation in Python [as I suggested in [the comment above](https://github.com/srl-labs/containerlab/issues/824#issuecomment-1073065513)]. Not sure why you want to blame our redhat friends ;-)

How long did you wait for your script to complete?
I just timed closerange() for ~1 mil values (our new max), which produced ~0.5s delay, and 10 mil values, which increased completion time linearly to about ~5 seconds.

```
root@4c774d3d1f39:/# time ./test2.sh 
real	0m0.517s
user	0m0.322s
sys	0m0.193s
root@4c774d3d1f39:/# time ./test3.sh 
real	0m5.021s
user	0m3.253s
sys	0m1.743s
```

So with 2^30 (1+ billion) descriptors and linear scale we are talking about 8-10 minutes on my system.
